tls: tidy up cockpit-certificate-helper, move "cleanup" from javscript

allisonkarlitskaya opened the pull request:
When we were originally creating this file, it was a very direct copy of the shellscript that was being assembled in-place by browser-side javascript code.  It was never meant to be a proper script.

Perform some general cleaning up, as well as splitting some of the code into functions.  Introduce the use of a permissions-restricted working directory (which will be cleaned up on exit).  Step up our automake game a little, and support installations not in /etc.

Finally: remove the +x bit.  It's meaningless on .in files; the script will be set +x on installation.

also:

Dismantle another shell-script-via-javascript instance for cleaning up the certificate and keytab on leaving an IPA domain and move the functionality to cockpit-certificate-helper.

While we're at it, make sure the .key file is also removed.


 - [x] #15602

martinpitt commented:
#15602 landed, so this can (and must) be rebased.

martinpitt commented:
Force-pushed a clean rebase, so that CI can run.

martinpitt commented:
Breaks `TestIPA.testQualifiedUsers` everywhere

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -2,51 +2,88 @@
-# use a temporary keytab to avoid interfering with the system one
-export KRB5CCNAME=/run/cockpit/keytab-setup
+install_cert() {
+    local DESTINATION="${COCKPIT_WS_CERTS_D}/$1"
```
Style nitpick: global vars should be UPPER_CASE, local vars lower_case IMHO. (opposite problem with `$prefix`)

allisonkarlitskaya replied:
`$prefix` is there to deal with the fact that the expansion for `@sysconfdir@` ends up containing, literally, `${prefix}`, so we don't have a lot of room to fix that.

In one of my attempts to do variable expansion in the manifests from webpack I had some tweaks to configure.ac to set an absolute version of @libexecdir@, and we could do a similar trick here.  Or we could just suffer with it as it is.

I was on the edge of adding a comment, and I think I will now...

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -2,51 +2,88 @@
 
 set -eu
```
I'm fairly sure we can use `set -f` here as wel, just in case the cert names contain funny characters.

allisonkarlitskaya replied:
Is there anywhere that isn't properly quoted that risks expansion?

I'd prefer to do this locally correct than to set this global flag...

martinpitt replied:
No, I didn't see insufficient quoting. It's more like belt-and-suspenders - i.e. we don't need glob expansion, so it could just as well disable it.

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -2,51 +2,88 @@
+    mkdir -p "${COCKPIT_RUNTIME_DIR}"
+    WORKDIR="${COCKPIT_RUNTIME_DIR}/certificate-helper"
+    mkdir -m 700 "${WORKDIR}" # we expect that not to have existed
+    trap 'rm -rf "${WORKDIR}"' EXIT
```
I wonder how that works, given that WORKDIR is the pwd then? doesn't that need to `cd ..` (or someplace else) first?

Please also add `INT QUIT PIPE`, to also clean up on some common errors.

allisonkarlitskaya replied:
I think there's no trouble with removing the current directory: particularly when you're right about to exit.

allisonkarlitskaya replied:
Also: adding `INT QUIT PIPE` here doesn't really do the right thing: it seems like it causes those signals to be responded to by removing the working directory (instead of quitting the shell).

`set -e`, however, seems to cause the shell to quit immediately anyway (including with the `EXIT` trap).  *not* trapping `INT` causes the `EXIT` trap *not* to run, though.

So it seems like we probably need to set INT QUIT PIPE, etc. to `''` (no action) and keep the `EXIT` trap?  This is confusing and awful :rofl:

allisonkarlitskaya replied:
I think the best way is to do something like this:

```sh
    trap 'exit' INT QUIT PIPE TERM
    trap 'rm -rf "${WORKDIR}"' EXIT
```

martinpitt replied:
Heh, that's not actually too bad then :)

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -2,51 +2,88 @@
+            ;;
+        *)
+            echo 'unknown subcommand'
+            ;;
```
missing `exit 1`

allisonkarlitskaya replied:
ok

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -26,6 +26,65 @@ install_key() {
+
+    # We renew certificates up to 30 days before expiry, so give ourselves a
+    # year, plus 30 days.  The maximum is variously mentioned to be 397 or 398.
+    DAYS=395
```
Wouldn't that make more sense a the top of the file, as a global constant? It's a bit weird that other functions access this and it looks like a global var, but isn't.

allisonkarlitskaya replied:
I actually want to mark this `local`.  `local` in shell is a bit weird: it contains the scope of the variable to the defining function, but called functions can still access it, and indeed, just like the names of the files to create, this variable is used in common to both of the self-sign cases, but not the IPA case, so it belongs here.

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -26,6 +26,65 @@ install_key() {
+        --hostname "${HOSTNAME}" \
+        --organization "${MACHINE_ID}" \
+        --subject-alt-name localhost \
+        --subject-alt-name IP:127.0.0.1/255.255.255.255
```
While we are at it, IPv6 ::1/128 too?

allisonkarlitskaya replied:
This seems to be non-trivial.  It's not entirely clear what the correct format for IPv6 is: different sources have conflicting answers.  Let's look at it in a separate PR.

martinpitt replied:
Ack.

martinpitt reviewed (changes requested):
Thanks!

allisonkarlitskaya commented:
Addressed all review comments and fixed the test failure, but now we have a new issue:

There's code in `cleanup_ws_credentials()` in `pkg/systemd/overview-cards/realmd-operations.js` that does the reverse operation.  It looks like this:

```javscript
                    var script = 'set -eu; [ $(id -u) = 0 ] || exit 0; ';

                    // clean up keytab
                    script += '[ ! -e /etc/cockpit/krb5.keytab ] || ipa-rmkeytab -k /etc/cockpit/krb5.keytab -p ' +
                    '"HTTP/$(hostname -f)@' + kerberos.RealmName + '"; ';

                    // clean up certificate
                    script += 'ipa-getcert stop-tracking -f /run/cockpit/ipa.crt -k /run/cockpit/ipa.key; ' +
                          'rm -f /etc/cockpit/ws-certs.d/10-ipa.cert; ';
```

...which is obviously no longer correct, but which also obviously should be also moved to be part of the certificate-helper script.

This will need some more work.

allisonkarlitskaya commented:
So, it seems that `getcert` supports tracking certificates for expiry and automatically renewing them.  That sounds wonderful, but it's kinda useless to us because of the SELinux problems requiring us to move the certificates into place from the temporary work directory...

Is there some other place we could install the certificate to, that SELinux won't get upset about, and use symlinks or something?

Otherwise, I suggest we disable the tracking feature.  That can be done with `--no-renew` but it leaves the open question about what to do with existing systems that have the tracking enabled, for a file in `/run` which doesn't exist anymore.

@mvollmer @martinpitt would love some input here :)

martinpitt commented:
@allisonkarlitskaya: As this feature does not work right now, my recommendation is to use --no-renew and create a jira card for revisiting "proper" renewal in the future. Next quarter I want to look at moving cockpit's SELinux policy from the central one into our own project, after that it may be simpler/possible to actually do the renewal.

allisonkarlitskaya commented:
In local testing, this is failing:

```
        # should have stopped cert tracking
        wait(lambda: "status:" not in m.execute("ipa-getcert list"))
```

because the output shows

```
Number of certificates and requests being tracked: 1.
Request ID '20210404205534':
	status: MONITORING
	stuck: no
	key pair storage: type=FILE,location='/run/cockpit/certificate-helper/10-ipa.key'
	certificate: type=FILE,location='/run/cockpit/certificate-helper/10-ipa.cert'
...
```

despite `--no-renew` being used.

I guess we'll have to look into this a bit further...

allisonkarlitskaya commented:
Okay.  My preferred solution here is to keep the status quo: certificate tracking for a file which doesn't exist.

Let's address this issue in a separate PR later.

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -56,6 +56,17 @@ cmd_ipa_request() {
+
+        # clean up certificate
+        ipa-getcert stop-tracking -f 10-ipa.cert -k 10-ipa.key
+        rm -f "${COCKPIT_WS_CERTS_D}"/10-ipa.*
```
Can you please spell out .key and .cert here instead of the glob? We don't want this to remove backup files, nor globbing.

allisonkarlitskaya replied:
OK.

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -114,6 +173,9 @@ main() {
 
     # Dispatch subcommand
     case "$1" in
+        selfsign)
+            cmd_selfsign
```
Let's please move this commit into a future PR that actually uses this new verb. Dead code is never nice..

allisonkarlitskaya replied:
Ya.  I was thinking that's a good idea.  And rename this PR, then.

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -28,6 +28,65 @@ install_key() {
+
+cmd_selfsign() {
+    # Common variables used by both methods
+    local MACHINE_ID="$(tr -c '[:xdigit:][:space:]' '.' < /etc/machine-id)"
```
What does that do? aside from the terminating LF, the file isn't supposed to contain anything other than hex digits anyway (and not even spaces)

allisonkarlitskaya replied:
It's a fairly direct port of the old C code:

```
  if (g_file_get_contents ("/etc/machine-id", &machine_id, NULL, NULL))
    g_strstrip (g_strcanon (machine_id, HEX, ' ')); /* both functions modify machine_id in-place */
```

the variable assignment will absorb the whitespace (which we allow to pass through unmodified, avoiding to change `\n` to `.`, for example).  any non-hex characters are canonicalised...

I clearly misread the original though and saw a `.` there and not a space ` `.  Will adjust that.

allisonkarlitskaya replied:
→ #15608

martinpitt reviewed (changes requested):
Thanks!

martinpitt commented on src/tls/cockpit-certificate-helper.in:
```diff
@@ -56,6 +56,17 @@ cmd_ipa_request() {
+
+        # clean up certificate
+        ipa-getcert stop-tracking -f 10-ipa.cert -k 10-ipa.key
+        rm -f "${COCKPIT_WS_CERTS_D}"/10-ipa.{cert,key}
```
Bashism, rather spell this out. This may possibly be the reason for the Debian/Ubuntu test failure? (I didn't look closely)

allisonkarlitskaya replied:
after trying about 10 other ways to avoid repeating the variable expansion, i couldn't find anything that was truly nicer...

martinpitt reviewed (approved):
Thanks!
